### PR TITLE
Bulk Promo Code Collision Fix

### PR DESCRIPTION
**Pull request by philbirt**
* Check in-memory promotion codes for duplication when building bulk codes

### CODE REVIEW

**Comment by jordan-brough** [comment]
:+1:

**Comment by philbirt** [comment]
@cbrunsdon @gmacdougall @jhawthorn Mind taking a look? Thanks!

**Comment by gmacdougall** [comment]
Looking now...still trying to wrap my head around things.

File: core/spec/models/spree/promotion_spec.rb [hunk 1]
```diff
@@ -598,20 +598,32 @@
...
       it "builds codes with the same base prefix" do
-        expect(promotion.codes.map(&:value)).to all(match(/\Aabc_/))
+        subject
+        expect(promotion.codes.map(&:value)).to match_array(["abc_NCCGRT", "abc_KZWBAR"])
+      end
+
+      context "there is a collision with the random codes generated" do
+        it "will resolve the collision" do
+          # Generate the same code twice, then generate a new code
+          promotion.should_receive(:code_with_randomness).twice.and_return("duplicate_code")
```
Comment (gmacdougall) [inline core/spec/models/spree/promotion_spec.rb]: Stubbing a method on the class under test is not ideal. Can you come up with a better method of doing this?

File: core/app/models/spree/promotion.rb [hunk 2]
```diff
@@ -218,14 +219,18 @@ def usage_count_for(promotable)
...
-        build_code_with_base(base_code)
+      if Spree::PromotionCode.where(value: random_code).exists? || codes.any? {|c| c.value == random_code }
+        build_code_with_base(base_code: base_code)
       else
-        codes.build(value: code_with_entropy)
+        codes.build(value: random_code)
       end
     end
+
+    def code_with_randomness(base_code:)
```
Comment (gmacdougall) [inline core/app/models/spree/promotion.rb]: Does this method need to be public?

File: core/spec/models/spree/promotion_spec.rb (hunk 1 above)
Comment (philbirt) [inline core/spec/models/spree/promotion_spec.rb]: Totally agreed on that notion. I'm not sure how I'll able to expose this particular case where the random generator spits out two of the same code without stubbing, though.

File: core/app/models/spree/promotion.rb (hunk 2 above)
Comment (philbirt) [inline core/app/models/spree/promotion.rb]: Not that I know of, is there a reason it should be?

Comment (athal7) [inline core/app/models/spree/promotion.rb]: it currently is

Comment (philbirt) [inline core/app/models/spree/promotion.rb]: Nope..? https://github.com/bonobos/spree/pull/290/files#diff-61724faa21c94b40f3ab9d29a665dc67R195

Comment (athal7) [inline core/app/models/spree/promotion.rb]: my mistake

Comment (gmacdougall) [inline core/app/models/spree/promotion.rb]: Sorry. That was my bad.

**Comment by gmacdougall** [comment]
:+1 looks much better

**Pull request merged**